UefiPayloadPkg/Library/BuildFdtLib/X86_BuildFdtLib.c

```diff
@@ -0,0 +1,971 @@
+                                              &mPciDevicePpi->PciIo,
+                                              (EFI_PCI_IO_PROTOCOL_WIDTH)EfiPciWidthUint16,
+                                              PCI_DEVICE_ID_OFFSET,
+                                              sizeof (PciData.Hdr.DeviceId),
+                                              &(PciData.Hdr.DeviceId)
+                                              );
+
+          Status = mPciDevicePpi->PciIo.Pci.Read (
+                                              &mPciDevicePpi->PciIo,
+                                              (EFI_PCI_IO_PROTOCOL_WIDTH)EfiPciWidthUint8,
+                                              PCI_REVISION_ID_OFFSET,
+                                              sizeof (PciData.Hdr.RevisionID),
+                                              &(PciData.Hdr.RevisionID)
+                                              );
+
+          Status = mPciDevicePpi->PciIo.Pci.Read (
+                                              &mPciDevicePpi->PciIo,
+                                              (EFI_PCI_IO_PROTOCOL_WIDTH)EfiPciWidthUint16,
+                                              PCI_SVID_OFFSET,
+                                              sizeof (PciData.Device.SubsystemVendorID),
+                                              &(PciData.Device.SubsystemVendorID)
+                                              );
+
+          Status = mPciDevicePpi->PciIo.Pci.Read (
+                                              &mPciDevicePpi->PciIo,
+                                              (EFI_PCI_IO_PROTOCOL_WIDTH)EfiPciWidthUint16,
+                                              PCI_SID_OFFSET,
+                                              sizeof (PciData.Device.SubsystemID),
+                                              &(PciData.Device.SubsystemID)
+                                              );
+        }
+
+        Data16 = CpuToFdt16 (PciData.Device.SubsystemID);
+        Status = FdtSetProp (Fdt, GmaNode, "subsystem-id", &Data16, sizeof (UINT16));
+
+        Data16 = CpuToFdt16 (PciData.Device.SubsystemVendorID);
+        Status = FdtSetProp (Fdt, GmaNode, "subsystem-vendor-id", &Data16, sizeof (UINT16));
+
+        Data16 = CpuToFdt16 (PciData.Hdr.RevisionID);
+        Status = FdtSetProp (Fdt, GmaNode, "revision-id", &Data16, sizeof (UINT16));
+
+        Data16 = CpuToFdt16 (PciData.Hdr.DeviceId);
+        Status = FdtSetProp (Fdt, GmaNode, "device-id", &Data16, sizeof (UINT16));
+
+        Data16 = CpuToFdt16 (PciData.Hdr.VendorId);
+        Status = FdtSetProp (Fdt, GmaNode, "vendor-id", &Data16, sizeof (UINT16));
+      }
+
+      DEBUG ((DEBUG_INFO, "BusBase %x, , PciRootBridgeInfo->RootBridge[Index].Bus.Base  :%x\n", BusBase, PciRootBridgeInfo->RootBridge[Index].Bus.Base));
+
+      if (PciRootBridgeInfo->RootBridge[Index].Bus.Base == BusBase) {
+        Status   = AsciiSPrint (eSPIStr, sizeof (eSPIStr), "isa@%X,%X", ESPI_DEV_NUM, ESPI_FUN_NUM);

```

Comment:
1. if SerialParent->IsIsaCompatible == TRUE, we will create serial port device node under isa parent node, Bus/Dev/Func of isa node should come from SerialParent->ParentDevicePcieBaseAddress.

2. if SerialParent->IsIsaCompatible == FALSE, we will create serial port device node under one of the pci-rb node basing on the parent bus number faling into which pci-rb range. parent bus number should come from SerialParent->ParentDevicePcieBaseAddress.

Reply:
Todo.

Reply:
resolved